### PR TITLE
Evaluating APM Overhead

### PR DESCRIPTION
**Pull request by mattsonjoshcf**
- [X ] Mini bio
Senior Application Support Engineer and 6 year Racker on the Critical Application Services team. Specializing in Java web applications
 - [ X] Desired date of publishing
11/14/16
 - [ X] Social media contact info
N/A
 - [ X] Redaction: (yes|no)
yes
 - [X ] Comments: (yes|no)
No

@catlook Please review.

### CODE REVIEW

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-d49b65fce7/).

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+comments: false
+author: Josh Mattson
+published: true
+authorIsRacker: true
+authorAvatar: 'http://www.gravatar.com/avatar/791617263c70278859e1b26c15d13eab'
+categories:
+    - DevOps
+---
+
+Application Performance Management tools can provide incredibly valuable insight in to the performance of your applications and ultimately your end users' experience.  This insight, however, does not come without its cost.  Because APM tools instrument code at runtime, there is always some level of performance overhead.  In contrast to some older APM tools, modern APM tools are designed to minimize the negative performance impact as much as possible to allow you to safely run them in production without your end users' experience suffering.  In this post we'll be evaluating the overhead introduced on an Adobe Experience Manager environment by two popular APM tools: [AppDynamics](http://www.appdynamics.com) and [New Relic](http://www.newrelic.com)
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: Application Performance Management (APM) tools ...
... insight into the  ...
... In this post, we'll evaluate the ...
... Adobe Experience Manager (AEM) environment ...

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+<!-- more -->
+
+### The Test
+
+In order to evaluate the performance impact from introducing these agents, we conducted a series of tests designed to minimize outside variables.  Our testing environment consisted of an AEM 6.2 application running on Rackspace's Managed Virtualization offering.  We utilized Adobe's [ToughDay](https://docs.adobe.com/docs/en/aem/6-2/develop/test/tough-day.html) tool to execute a simple test case that mimicked an authoring user uploading a large amount of digital assets in to the repository.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: To evaluate the performance impact of introducing ...
... We used Adobe's ...
... assets into the ...

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+To prevent any confounding variables such as AEM maintenance operations, workflow/rendition processing or others, the application environment was completely scrubbed and rebuilt from the ground up after every test utilizing the [Rackspace MAS](https://www.rackspace.com/digital/managed-application-services) automation infrastructure.  This ensured consistency and repeatability over our test cases.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: .... variables (such as AEM maintenance operations, workflow/rendition processing or others), we completely scrubbed the application environment and rebuilt it from the ground up after every test by using the ...

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+We first started by executing multiple runs of the test against the environment with no APM tool in place as a control.  Once the results of these tests had been recorded, the tests were repeated the same number of times with New Relic and AppDynamics Java APM agents in place.  With these numbers in hand, we were able to calculate the performance overhead from loading an APM tool.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: We started by ...
... recorded, we repeated the tests an identical number of times ....

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+### The Results
+
+Before we dive in to the results, it's important to note that the overhead of APM tools will vary depending on the application and current usage patterns.  While our tests can illustrate and give us a general idea of the performance overhead, these results may not match your particular environment.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: ... APM tools varies depending ...
... tests illustrate (and give us a general idea of) the performance  ...

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+![APM Overhead Test Results]({% asset_path 2016-11-14-Evaluating-APM-Overhead/apm_overhead_results.png %})
+
+Our control test case with no APM agents had a mean processing time of 262 seconds, whereas the tests utilizing APM tools had a mean processing time of 270 seconds.  Overall, the increase is minimal and on the lower end of New Relic's and AppDynamics' advertised performance overhead.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: ... seconds, and the tests using APM ...
... minimal, and it's on the ...

File: _posts/2016-11-14-Evaluating-APM-Overhead.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+### The Takeaway
+
+While it's true that APM tools *do* have a measurable impact on your application's performance, modern APM tools like New Relic and AppDynamics have all but eliminated this performance overhead.  There is still a trade off, but the amount of insight in to your application's performance these tools bring can help you to reduce time to resolution, identify code problems before they hit production, more effectively use your application's infrastructure and ultimately improve your end user experience.  With that in mind, choosing to instrument your application with an APM tool seems like a pretty clear choice.
```
Comment (catlook) [inline _posts/2016-11-14-Evaluating-APM-Overhead.md]: ... insight into your application's performance that these tools ...
... infrastructure, and ...
[ Ed note: I've never heard the phrase "to instrument an application". If it's legit, go for it.  If in doubt, maybe enhance the application instead? ]

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-cb2010cc2e/).

**Comment by mattsonjoshcf** [comment]
Edits committed, thanks!

**Comment by catlook** [comment]
Thanks, Josh!  I'll publish on Monday :)

**Pull request merged**